### PR TITLE
#624 Set flat priority only for services with traffic splitting

### PR DESCRIPTION
**Pull request by nastassia-dailidava**
Added setting: "zonesAllowingTrafficSplitting", so changes in a config would be made only for envoys in that zone
Fixed setting priority for traffic splitting endpoints, they will be duplicated with higher priorities

closes allegro-internal/flex-roadmap#624

[tests](https://docs.google.com/document/d/12izmBIDbSkeVLfrXltxjG1zpPkA2PAI8kURakwjwGi0/edit#heading=h.xsqndl5asmj2)

### CODE REVIEW

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/clusters/EnvoyClustersFactory.kt [hunk 1]
```diff
@@ -283,15 +283,11 @@ class EnvoyClustersFactory(
...
+        val allowed = clusterLoadAssignment != null &&
+            properties.loadBalancing.trafficSplitting.zonesAllowingTrafficSplitting.contains(currentZone)
+        return trafficSplitEnabled && allowed
     }
 
-    private fun hasEndpointsInZone(
-        clusterLoadAssignment: ClusterLoadAssignment?,
-        trafficSplitting: TrafficSplittingProperties
-    ) = clusterLoadAssignment?.endpointsList
-        ?.any { e -> trafficSplitting.zoneName == e.locality.zone && e.lbEndpointsCount > 0 } ?: false
```
Comment (KSmigielski) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/clusters/EnvoyClustersFactory.kt:293]: Why we no longer check if there are any endpoints?

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/SnapshotProperties.kt [hunk 2]
```diff
@@ -166,6 +166,7 @@ class TrafficSplittingProperties {
     var zoneName = ""
     var headerName = ""
     var weightsByService: Map<String, ZoneWeights> = mapOf()
+    var zonesAllowingTrafficSplitting = listOf<String>()
```
Comment (kozjan) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/SnapshotProperties.kt:169]: does defining it like this allow for using both `zonesAllowingTrafficSplitting` and `zones-allowing-traffic-splitting`? 
I see that the properties are used both ways in [config PRs](https://github.com/allegro-internal/alle-envoy-control-config/commit/d4d26a7ba87be935b43086ef659a8e2df8e95529#diff-438d442126cd51412e0df44a27f72c0aaec0690a0c93af109301c39e4411a9eeL130), other properties in [this PR](https://github.com/allegro/envoy-control/pull/414/files#diff-4c14cffe1498e7d00744dffea81106f21fce5b1b9c4d20524364006d0d949fa8R25)
Traffic splitting tests use [zonesAllowingTrafficSplitting](https://docs.google.com/document/d/12izmBIDbSkeVLfrXltxjG1zpPkA2PAI8kURakwjwGi0/edit#bookmark=id.9mju8widuelu), while docs have [zones-allowing-traffic-splitting](https://docs.google.com/document/d/1KVmEu3Cqi9E4TI4-ot8CFVyxjblTi69gQWiHaxjRYEI/edit#heading=h.fiidnpxsl1ge)

also pls add to `configuration.md` and while you are at it I think `envoy-control.envoy.snapshot.load-balancing.traffic-splitting.weights-by-service-properties` should be fixed to `envoy-control.envoy.snapshot.load-balancing.traffic-splitting.weights-by-service`

Comment (nastassia-dailidava) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/SnapshotProperties.kt:169]: so it's allowed both ways, in all aec configs mostly used kebab case, so I added examples to the docs similarly, in the tests I was just copying from properties so it's camel case, I fixed some properties in tests to make them consistent throughout the test, but I can fix to kebab case as well

Comment (nastassia-dailidava) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/SnapshotProperties.kt:169]: and fixed property in configuration.md

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/clusters/EnvoyClustersFactory.kt (hunk 1 above)
Comment (nastassia-dailidava) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/clusters/EnvoyClustersFactory.kt:293]: yea I think I had some justification when I was deleting it, but now don't recall why, so returned it back

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/clusters/EnvoyClustersFactory.kt [hunk 3]
```diff
@@ -283,14 +283,16 @@ class EnvoyClustersFactory(
     ): Boolean {
         val trafficSplitting = properties.loadBalancing.trafficSplitting
         val trafficSplitEnabled = trafficSplitting.weightsByService.containsKey(serviceName)
-        return trafficSplitEnabled && hasEndpointsInZone(clusterLoadAssignment, trafficSplitting)
+        val allowed = clusterLoadAssignment != null && hasEndpointsInZone(clusterLoadAssignment)
+        properties.loadBalancing.trafficSplitting.zonesAllowingTrafficSplitting.contains(currentZone)
```
[the remark this answers is not in the document]
Comment (kozjan) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/clusters/EnvoyClustersFactory.kt:287]: probably better to edit it by hand in IDE than committing suggestion, bc i'm not sure if whitespace is correct

[the remark this answers is not in the document]
Comment (nastassia-dailidava) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/clusters/EnvoyClustersFactory.kt]: true true, thx fixed

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/clusters/EnvoyClustersFactory.kt [hunk 4]
```diff
@@ -283,14 +283,16 @@ class EnvoyClustersFactory(
...
     }
 
     private fun hasEndpointsInZone(
-        clusterLoadAssignment: ClusterLoadAssignment?,
-        trafficSplitting: TrafficSplittingProperties
+        clusterLoadAssignment: ClusterLoadAssignment?
     ) = clusterLoadAssignment?.endpointsList
-        ?.any { e -> trafficSplitting.zoneName == e.locality.zone && e.lbEndpointsCount > 0 } ?: false
+        ?.any { e -> properties.loadBalancing.trafficSplitting.zoneName == e.locality.zone && e.lbEndpointsCount > 0 }
+        ?: false
```
[the remark this answers is not in the document]
Comment (nastassia-dailidava) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/clusters/EnvoyClustersFactory.kt:295]: hehe it was like this before, and I changed it, cause i thought it makes a little bit of confusion, cause this property is declared already and it's constant

[the remark this answers is not in the document]
Comment (nastassia-dailidava) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/clusters/EnvoyClustersFactory.kt:295]: well maybe passing zone would be better yea

File: docs/configuration.md [hunk 5]
```diff
@@ -138,7 +138,8 @@ Property
...
 **envoy-control.envoy.snapshot.load-balancing.traffic-splitting.zoneName**             | a zone to which traffic will be routed if traffic splitting is enabled                                                                                                                 | ""
-**envoy-control.envoy.snapshot.load-balancing.traffic-splitting.weights-by-service-properties.**             | a map that maps service name to a map [zoneName: weight]                                                                                                                               | empty map
+**envoy-control.envoy.snapshot.load-balancing.traffic-splitting.zones-allowing-traffic-splitting**             | a zone from which traffic should be splitted                                                                                                                                           | empty list
+**envoy-control.envoy.snapshot.load-balancing.traffic-splitting.weights-by-service.**             | a map that maps service name to a map [zoneName: weight]                                                                                                                               | empty map
```
Comment (kozjan) [inline docs/configuration.md:142]:
```suggestion
**envoy-control.envoy.snapshot.load-balancing.traffic-splitting.weights-by-service**             | a map that maps service name to a map [zoneName: weight]                                                                                                                               | empty map
```
let's kill this dot 💀

**Review by KSmigielski — APPROVED** [review]

**Pull request merged**